First version of the mechanics of contributing article, addresses #74 (#117)
InnerSourceCommons/InnerSourceLearningPath: contributor/04-mechanics-of-contributing.md

# Article

You are ready to start contributing to other teams projects/repos or look forward to reduce your blockers by not playing the big-cheese game but by collaborating? 
Please consider reading this section beforehand as it will give you practical advice and informs you about gotchas to remember when making an InnerSource contribution. This will hopefully enable you and the host team to have as pleasant an experience as possible setting the foundation for more contributions and great collaboration.  

This article is separated into the three steps you will likely experience

- Soliciting your contribution opportunity and preparing to work on it
- Actually creating the contribution
- Polishing and wrapping the gift nicely and presenting it to the host team.

If your contribution is larger, you'll possibly go through (some) these steps repeatedly as you iterate towards your common goal.
It is very likely that as you do this everything will feel more and more naturally - maybe you'll even wonder why you were doing anything else before.

## Preparing to work

### Lead times

> 

Reviewer: That's a great suggestion: Backing that up with research findings would be even better.

Even if you are coming back after a while the teams' ways and members might have changed - take your time as you would to catch up with a friend you haven't seen in a while and whom you are visiting now.

Thus: Please give yourself enough lead time.
Start early enough so that your work is available for you to leverage at the time you need it. 
It's better to add more slack time initially - you'll get a feeling about the turn-around times once you work with the host team.
If you end with too much there surely will be a great way to make use of the time, be it some work on your side, the quirky bug or improvement opportunity you found while making your initial contribution or possibly even a nice chat over coffee, beers in case of personal availability or whatever topic in case of distributed projects. 

[EDITOR NOTE: There was actually a paper on how trust is build up on remote/asynchronous collaboration. Maybe I'll find that again. 
The basic point is that aside of reliability in collaboration, continued presence the effect of water cooler discussions and food/drink consumption is quite impressive...

> ]  

Author: > Maybe that's an additional paragraph, not something to include in the "give enough lead time" part.

Yup, I guess so. Could you either ma PR against this branch out of it if you have a nice paragraph on it?

Especially in the case of larger distances between you and the host teams, e.g. multiple time zones difference as in San Francisco, USA (PDT) and Berlin, Germany (CEST), you will need to actively consider this and explicitly manage your and their expectations on reaction times.

> 

Reviewer: ### Building trust

InnerSource puts huge weight on written communication - in particular when it comes to project decisions. Does that imply that in-person communication is forbidden?

Clearly not: Where written communication shines when it comes to archiving and searchability, in-person communication shines when it comes to communication bandwidth. Try to make time to meet the people behind the names. If possible, try to meet them over your favorite beverage or some food. When you're able to hear people speak, when you know their idiosyncrasies remote collaboration will become easier. 

[EDITOR NOTE: There was actually a paper on how trust is build up on remote/asynchronous collaboration. Maybe I'll find that again. 
The basic point is that aside of reliability in collaboration, continued presence the effect of water cooler discussions and food/drink consumption is quite impressive...
]

### Avoiding rejection

You've got this really large feature or larger rewrite in mind that would be totally great and add real value that you would like to donate?
Excellent!
Wouldn't it be horrible if all your work would be wasted because the host team is maybe already building something very similar, is planning to depreciate the part of the software you are trying to extend or even does not see what you are proposing to be a fit for their project? 
This challenge is a frequent challenge and many team relationships suffered from not handling the fit in advance.

> Make yourself and the host team happy (and possibly even save some work) by getting agreement from the host team on the user/technical design of the contribution _before_ submitting a pull request.

Author: Oh yes, very important difference indeed.

> You'll have to understand how the host team would like you to reach out for this - sometimes they have regular design discussions in person or via chat/voice/video conferences, some hand in design discussion / extension proposal documents for collective discussion and review, some work with feature requests, etc. 

Reviewer: This is the Apache person in me speaking: I'd be careful with even mentioning that a host team relies on regular design discussions in person or via chat/voice/video conferences. Great Inner Source projects favor written over verbal advise - this should also include design decisions.

I'd split that sentence into multiple pieces:

- How to figure out, where the project is heading (contributing.md should tell you through which channels the host team discusses and decides on design, roadmap etc.)
- Mention that while there may be in person discussion through various channels, contributing.md will tell you where to find summaries of these discussions and which channels to use to participate in these discussions in an asynchronous way.

I admit that training contributors to go look for these resources and teaching them to demand these resources is a hack towards making host teams operate in a more remote friendly, async friendly way. That in turn should help with creating passive documentation of the project direction. It should also help with answering the question of "why did we choose that implementation path" a couple months or years later.

Author: You are very right here.
I'll try to rephrase things here. I'd also add that having regular non-written conversations might be totally ok, however that time-zone-based participation limitations need to be considered for inclusivity and that the final goal needs to be having everything written & linkable/searchable.

Side note: an interesting approach are some of the meetings that OpenZFS holds are recorded and put online on Youtube - generating possibly searchable subtitles. Pure text is easier though...

> Ask a trusted committer about how to best discuss your proposal.  

Reviewer: I wouldn't encourage contributors to go the 1:1 conversation path. They should be directed to the project communication channel that is watched over by a group of trusted committers. 

Where am I coming from with that advise:

http://people.apache.org/~hossman/#private_q

Author: I agree, would you care to make commit suggestion to change that line to that version? (Otherwise I'll do that)

This type of high-level, up-front agreement early will save time in rework or rejection of your pull request down the road.

## Creating the pull request

### Communication and unblocking yourself

> Great, you've made yourself familiar with the host teams approach and they are looking forward to receive your pull request.

Reviewer: ```suggestion
Great, you've made yourself familiar with the host team's approach and they are looking forward to receive your pull request.
```

Which gotchas are there waiting for you now?

First, you'll be in less direct contact with them and second you aren't expected to be as knowledgeable and proficient as you might be on the full-time projects that your team owns.
How can you now deal with this the best?

Try to peruse their documentation, the conversation archives and code artifacts from the host team to unblock yourself - this is pretty much similar to the situation you and likely most people find yourself in when using one of the popular OSS projects.

> However, this time you've got the advantage to be able to directly ask the host team if, even after trying to unblock yourself, things went nowhere. (That's an approach that works with many OSS projects as well!)

Reviewer: ```suggestion
Much like in Open Source projects, make use of the advantage to be able to directly ask the host team if, even after trying to unblock yourself, things went nowhere. The questions you ask and the answers you receive will help others coming after you solve the same issues - so make sure that your communication ends up in a searchable archive that is closely linked to the project itself.
```

It doesn't do either side any good for you to spin forever on a problem that could be resolved in a few-minute conversation with someone more knowledgeable about the project.
It's OK to ask for help.

There's one key difference though, bringing advantage to you and other people in the future: 

> In almost all cases you should prefer the projects' official communication channels - this can be a mailing list, a chat room, the issue tracker or something similar depending on the purpose of having a more synchronous or asynchronous way of interacting or the varying needs for structure in the communication.

Reviewer: I would bring that up much earlier, already in the part where you talk about working on the design and deciding whether or not to submit the pull request.

Reviewer: Likely I would give more advise on which question to ask where here. Check out https://www.youtube.com/watch?v=HPjYa_O6yEk for what I mean

Author: > I would bring that up much earlier, already in the part where you talk about working on the design and deciding whether or not to submit the pull request.

I've committed your suggestion w.r.t. to this and will see if I find something more to add after watching the video.

> All of those usually have in common that they are text-based, archived, searchable and linkable - this means your question and the answer will be written down.

Reviewer: ```suggestion
All of those usually have in common that they are text-based, archived, searchable and come with stable links - this means your question and the answer will be written down, references you link in those answers will also be kept reachable.
```

> This way you could benefit from this passively documented knowledge on your search and will help future contributors to have the same advantage.

Reviewer: I would mention that having the answers in writing also means that you can easily link to them in your official documentation should you later need them, instead of having to write them down again. 

See https://www.apache.org/dev/committers.html and search for "volunteeritis" on the page to find a link to a mail that was deemed important enough to be included in the new committer documentation.

Reviewer 2: If we put in this extra explanation then I think that we should remove the following sentence.  Only need one additional sentence of elaboration, I'd say.

Author: Line 108 serves as an example for why the approach in line 106 might be great.
I would think that the extra example provide above is somewhat orthogonal to line 108 but don't have a real firm opinion here. What does @MaineC think here?  https://github.com/InnerSourceCommons/InnerSourceLearningPath/pull/117/files#r328171621

Imagine what would be needed for realizing this benefit by just walking by or calling a trusted committer on the project or even in a group call.    

Did you on your way of trying to unblock yourself find that you were missing documentation or found parts of it to be wrong or out-of-date and had to figure this out for yourself now?
Often projects are happy to receive additions, updates or corrections for their existing documentation as well - you've just found another opportunity to contribute! 
(Or just politely provide them with feedback on your experience and what would have helped you.)

### Crafting the code

We all have our preferences and opinions on code style, indentation, etc. 
The home that the host team lives in will have them as well, individual wallpaper, decoration and hangings incl. 
Try to adapt and match those preferences even if it's not what you would normally do and even if it is not specified in the projects' `CONTRIBUTING.md`.
If you are unsure you can always ask politely, nevertheless a guest contribution for a feature or bug fix is not the time to introduce a new way of structuring or formatting project code.

> ## Submitting the pull request

Reviewer: I'm missing one aspect here: Release early, release often - in order to avoid running in the wrong direction, it can help to submit a draft pull request, ask for early feedback and go from there. There's even a GitHub feature for that.

This advise comes with two caveats:

People in corporate settings usually are unwilling to share non-polished work, exposing themselves to criticism early on. Likely that's particularly true in environments that aren't tolerant wrt. mistakes and weaknesses.

People in corporate settings in a host team may be unwilling to review draft work - and reviewing that over and over until it's done. Most likely this applies to those who are not used to work in tight iteration loops.

Author: Important addition!
Especially the case of willingness to share non-polished work or work in the open from very early on due to fear of retaliation or ridicule or exhaustion of help offers or patience.

Not instantly sure how to write that though. Let's see what I can come up with later.

Author: Just took a look through the text again and I would think that [this](https://github.com/InnerSourceCommons/InnerSourceLearningPath/pull/117/files#diff-7edd0fadfb4f25c945bdeca937c16632R148-R150) addresses some of this.
I'll add a few more lines to it.

You've completed all the essential work, figured out all the quirks of the problem and the project you are contributing to, the time you've planned for the new feature to be used comes nearer and you want to make sure your contribution gets merged in as fast and smooth as possible?

Here's what you can do to make reviewing and merging as easy as possible for the the trusted committer and the host team.
(This might actually be pretty similar to what you might already be doing on your own project to get your changes accepted - if that's the case: great, this is going to come natural to you!)

### Testing and automation

> The basic point here is to enable the trusted committer to validate the contribution without your presence and to ensure easy maintainability.

Reviewer: If you need a real story for that:

I once added a feature that I needed at work to an open source project where I was committer - only to have it removed in the next performance optimzation round. It didn't come with a test, so on removal nobody was notified that it was actually needed - nor what it would have been needed for.

Author: Valid point, do you see a point where this example could be added while keeping the reading flow? 
I tried and couldn't really decide upon a place to put it.
Would be very about a commit suggestion here.

Author: Added a short commit suggestion incorporating your story. Take a look @MaineC 

```suggestion
The basic point here is to enable the trusted committer to validate the contribution without your presence and to ensure easy maintainability.
Imagine you've built a feature or handling of an unsolvable quirk or important performance tweak whose code is not entirely obvious or might even look hacky / wrong at the first glance. 
If you have covered this with a test - and ideally have shed some words on the rationale behind it in a comment - a future editor will get reminded about the purpose of the code and the test(s) will ensure that the value your code realizes will be kept, even in new implementations.
```

Author: Offline comment from @MaineC: wfm

To achieve this do the following: 

- Add tests for your code contribution so that validating the function of your contribution by others works well, even after some time when you work in other projects or might have stopped to contribute to this project.

>   - Often projects will use automated checks against pull request using those tests and the level of code coverage. Try to meet the criteria these test enforce.

Reviewer: ```suggestion
  - Often projects will use automated checks against pull requests using those tests and the level of code coverage. Try to meet the criteria these tests enforce.
```

> - Many projects will provide project build and validations scripts to enable you to locally test your changes. 

Reviewer: ```suggestion
- Many projects will provide project build and validation scripts that enable you to locally test your changes. 
```

  - Use those to ensure that your contribution works as well as possible before opening a pull request. 
  - Having to review defect pull requests with easy-to-fix errors often bugs trusted committers. They will not fix your code but ask you to do so. This might create more round-trips and slow the merge if there's unnecessary errors.

>   - No one's perfect though. Try your best, use the prepared validation scripts if there are any and give it your best shot with a pull request!

Reviewer: I would add that tests you broke could also be explicitly be highlighted in the pull request comment to ask for help understanding what happened and how to fix it.

Author: ```suggestion
  - No one's perfect though. Try your best, use the prepared validation scripts if there are any and give it your best shot with a pull request!
  - I your pull request keeps breaking tests and you can't find out why after giving it your best shot try to highlight those tests in the pull request comment, illustrate your current understanding of the problem and ask for help on it.  
```

Added a small sentence on this.

- Don't forget your own project that triggered your contribution in the first place. Build a modified build of the shared project with your changes and try it out in your own project that consumes it.

### Documentation and reviewability

> You'll want to ensure that your pull request includes any documentation updates are relevant to your changes. 

Reviewer: ```suggestion
You'll want to ensure that your pull request includes any documentation updates that are relevant to your changes. 
```

Should the documentation live in a different place, make sure you add them there and link to them in the pull request.

To make the actual code review as easy as possible for the trusted committer or other persons reviewing it, try to follow these hints:

- Be sure that your pull request includes just the relevant changes for the issue you're completing.
- Try to avoid super-large commits, commits with unclear commit messages, gazillions of files, incoherent changes (e.g. touching multiple topics).
- Provide a clear description what this pull request changes, why it does so and which issue and possible design documents (if there were any) it refers to.
- If there is anything uncommon or unexpected in the pull request highlight it and provide an explanation. This will make it easier to reason about and solve potential blocking questions the reviewer might have during the review. 
  - The same goes for place where you were unsure in the implementation or your approach - highlight it and ask for insight.

>   - Be civil and expect civility from the trusted committers' review.   

Reviewer: Instead of a simple "be civil", maybe something like "expect best intentions" which I believe goes a step further?

Author: I think that's very important too, however (aside of the mild play on words) was aiming for "don't accept trusted committers being a jerk".
Thus: 
- Try to assume the best intentions from both sides
- Don't let people be a jerk to you / set some reasonable limits (especially since some cases of people being jerks have been reported from some corners of OSS-land I heard ;) and office politics as well as personal animosities might also lead to comparable effects.)

I'm not sure how to combine both things best though.

Reviewer: Ok - that's where you're coming from. If that is the goal, I think your original text captures it well.

Author: Good, closing this discussion then.

> - Making pull requests too broad and large makes them more difficult to review and longer before they're accepted.

Reviewer: ```suggestion
- Making pull requests too broad and large makes them more difficult to review so it will take much longer before they're accepted.
```

>   - If you have a larger feature you are contributing it often helps to split it in multiple pull request that get submitted, reviewed and accepted sequentially.

Reviewer: ```suggestion
  - If you have a larger feature you are contributing it often helps to split it in multiple pull requests that get submitted, reviewed and accepted sequentially.
```

>   - You can still bind them together with an issue the are referring to.  

Reviewer: ```suggestion
  - You can still bind them together with an issue that you are referring to.  
```

In addition to the technical side, you may be faced with differences in communication, think code-reviews.
Even if you are coming back after a while the teams' ways and members might have changed.
Take your time as you would to catch up with a friend you haven't seen in a while and whom you are visiting now.

Give yourself enough lead time.
Start early enough so that your work is available for you to leverage at the time you need it. 
It's better to add more slack time initially - you'll get a feeling about the turn-around times once you work with the host team.
Often, you will notice a reduction in turn-around time per host team after making a few successful contributions to that host team.
This is effect can be observed with Open Source as well, you can read more about it [here](https://doi.org/10.1109/MS.2013.95). 

[EDITOR NOTE: There was actually a paper on how trust is build up on remote/asynchronous collaboration. Maybe I'll find that again. 
The basic point is that aside of reliability in collaboration, continued presence the effect of water cooler discussions and food/drink consumption is quite impressive...
]  

> Especially in the case of larger distances between you and the host teams, e.g. multiple time zones difference as in San Francisco, USA (PDT) and Berlin, Germany (CEST), you will need to actively consider this and explicitly manage your and their expectations on reaction times.

Author: Added a paragraph on expectation management turned around in the way @MaineC suggested. What do you think?

```suggestion
### Expectation management
In your classic teams everyone had an idea of the expectable lead times.
Within an InnerSource context this might not be the case, either due to large time-zone differences (e.g. Seattle, USA with PDT vs Berlin, Germany with CEST) or you not being available full-time as with your original team even if they are in the same physical location as you are.
Thus, to prevent frustration on both sides, impatience and other non-trust-building effects you'll need to explicitly do expectations managemnet with regards to your expected reaction times.
One approach be to just quickly react with a "I'll look into it, I won't get to it in the next few days though" to a Trusted Committers feedback if you know that you'll only be able to come back to them in a few days.
Ideally you can provide them with a rough estimate when you likely will have time to take a look at their input.
Doing so builds trust by reliability even over non-physical contact, longer distance or otherwise asynchronous media.
Established trust will allow you overcome uncertainty bumps in the collaborative road ahead of you. 
```

Reviewer 2: LGTM

Reviewer: LGTM

This type of high-level, up-front agreement early will save time in rework or rejection of your pull request down the road.

## Creating the pull request

### Communication and unblocking yourself

Great, you've made yourself familiar with the host team's approach and they are looking forward to receive your pull request.
Which gotchas are there waiting for you now?

First, you'll be in less direct contact with them and second you aren't expected to be as knowledgeable and proficient as you might be on the full-time projects that your team owns.
How can you now deal with this the best?

Try to peruse their documentation, the conversation archives and code artifacts from the host team to unblock yourself.
This is similar to the situation you and likely most people find yourself in when using one of the popular OSS projects.

Much like in Open Source projects, ask the host team if things are going nowhere even after trying to unblock yourself.
The questions you ask and the answers you receive will help others coming after you solve the same issues.

> Make sure that your communication ends up in a searchable archive that is closely linked to the project itself.

Author: > > make sure that your communication ends up in a searchable archive
> 
> I think that this is the responsibility of the Host team instead of the Contributor. The Host team should encourage questions to go to a location where they can be searched later. The Contributor should just ask questions wherever the Host team encourages them to ask them.

I tend to agree here and would, from experience, argue that it could be acceptable to - very politely - ask questions even where they are not explicitly encouraged. Sometimes something just came into being by sheer coincidence and stuck (just as e.g. locations for content/questions). 
I've got a case of an established (Australian vendor) wiki for documentation that suffered from massive [vogonism](https://wiki.haskell.org/Vogonism) and outdated-on-write and an external suggestion to include documentation into the repo and update via PR&review that was happily accepted as a bit of anecdotal evidence.

> Make sure that your communication ends up in a searchable archive that is closely linked to the project itself.

Reviewer: I'd like to second what @lenucksi states here: While ultimately responsibility lies with the host team, having an educated contributor that knows how things are supposed to be done can help improve the host project (and avoid surprises about how things are being run on the contributor side)

> Make sure that your communication ends up in a searchable archive that is closely linked to the project itself.

Reviewer 2: That's  true.

> Make sure that your communication ends up in a searchable archive that is closely linked to the project itself.

Author: Added as suggestion to add exactly this discussion's outcome. Feel free to +1/-1 / change the commit suggestion. Otherwise I'd close this discussion.  

In case someone is interested in a more academic discussion or "organized anarchies" and their effects, this is actually a nice paper: https://www.jstor.org/stable/2392088 ("A Garbage Can Model of Organizational Choice", Michael D. Cohen, James G. March and Johan P. Olsen)

```suggestion
Make sure that your communication ends up in a searchable archive that is closely linked to the project itself.
Should you see easy improvement opportunities to reach said goal if it is not reached yet, you could try to - very politely - suggest an improvement to your host team.
Sometimes the status quo arises from pure happenstance and stays that way because no-one had a different idea or cared enough.
Suggestions for improvement might be welcome in such cases.
```

> Make sure that your communication ends up in a searchable archive that is closely linked to the project itself.

Reviewer 2: I'd just leave the text as-is, but we can see what @MaineC thinks.

> Make sure that your communication ends up in a searchable archive that is closely linked to the project itself.

Author: Leaving it like that then.

To make the actual code review as easy as possible for the trusted committer or other persons reviewing it, try to follow these hints:

- Be sure that your pull request includes just the relevant changes for the issue you're completing.
- Try to avoid super-large commits, commits with unclear commit messages, gazillions of files, incoherent changes (e.g. touching multiple topics).
- Provide a clear description what this pull request changes, why it does so and which issue and possible design documents (if there were any) it refers to.
- If there is anything uncommon or unexpected in the pull request highlight it and provide an explanation. This will make it easier to reason about and solve potential blocking questions the reviewer might have during the review. 
  - The same goes for place where you were unsure in the implementation or your approach - highlight it and ask for insight.
  - Be civil and expect civility from the trusted committers' review.   
- Making pull requests too broad and large makes them more difficult to review so it will take much longer before they're accepted.

>   - If you have a larger feature you are contributing it often helps to split it in multiple pull requests that get submitted, reviewed and accepted sequentially.

Author: ```suggestion
  - If you have a larger feature you are contributing it often helps to split it in multiple pull requests that get submitted, reviewed and accepted sequentially.
    - Some tools also have Draft / WIP pull request functionality that you can use to explicitly mark unfinished and non-polished work and still get early feedback from your host team's trusted committers.
    - This allows you to ensure you are going down a path that your host team is happy to merge once it's done, adhering to the "release early, release often" idea in a way.
    - The host teams' responsibility is to create an atmosphere where sharing and discussing not-totally-polished work is possible and welcome. If you can't fail safe, you can't innovate and collaborate becomes very hard.  
    - Try to balance between asking for review early and providing meaningful changes to review.
```

Added a few additional sentences based on this [comment](https://github.com/InnerSourceCommons/InnerSourceLearningPath/pull/117/files#r315655318). 
Please take a look @MaineC.

Reviewer: LGTM

Reviewer 2: @MaineC can you commit the suggestion?  I can't :(

Author: I can.

Reviewer 2: Great!

## Preparing to work

### Lead times

One key difference is the turn-around time. 
With every first time contribution you are coming to a new (host) team.
As a result, you'll need to get to know their code-base and the technology used, but also their preferred development environment (think test framework, build system).
Even in cases where this type of tooling is standardized each team will have developed some individual peculiarities.
In addition to the technical side, you may be faced with differences in communication, think code-reviews.
Even if you are coming back after a while the teams' ways and members might have changed.
Take your time as you would to catch up with a friend you haven't seen in a while and whom you are visiting now.

Give yourself enough lead time.
Start early enough so that your work is available for you to leverage at the time you need it. 
It's better to add more slack time initially - you'll get a feeling about the turn-around times once you work with the host team.
Often, you will notice a reduction in turn-around time per host team after making a few successful contributions to that host team.

> This is effect can be observed with Open Source as well, you can read more about it [here](https://doi.org/10.1109/MS.2013.95). 

Author: ```suggestion
This is effect can be observed with Open Source as well, you can read more about it [here](https://doi.org/10.1109/MS.2013.95). 
Unfortunately, this paper access restricted. 
Sometimes your employer has a subscription enabling access, otherwise public university libraries often allow access for guests, too.
```

Reviewer 2: Since many readers will not have access to this content can we move it to some sort of "references" section at the bottom instead of having it inline?

Reviewer 2: Thank you!

Make sure that your communication ends up in a searchable archive that is closely linked to the project itself.
It doesn't do either side any good for you to spin forever on a problem that could be resolved in a few-minute conversation with someone more knowledgeable about the project.
It's OK to ask for help.

There's one key difference though, bringing advantage to you and other people in the future: 
In almost all cases you should prefer the projects' official communication channels - this can be a mailing list, a chat room, the issue tracker or something similar depending on the purpose of having a more synchronous or asynchronous way of interacting or the varying needs for structure in the communication.
All of those usually have in common that they are text-based, archived, searchable and come with stable links - this means your question and the answer will be written down, references you link in those answers will also be kept reachable.
This way you could benefit from this passively documented knowledge on your search and will help future contributors to have the same advantage.

> Imagine what would be needed for realizing this benefit by just walking by or calling a trusted committer on the project or even in a group call.    

Author: Removing this line as suggested 
 by @rrrutledge  [here](https://github.com/InnerSourceCommons/InnerSourceLearningPath/pull/117/files#r324852819)

```suggestion
```

Reviewer 2: 👍 Thank you!

Reviewer 2: Please commit the suggestion?

Give yourself enough lead time.
Start early enough so that your work is available for you to leverage at the time you need it. 
It's better to add more slack time initially - you'll get a feeling about the turn-around times once you work with the host team.
Often, you will notice a reduction in turn-around time per host team after making a few successful contributions to that host team.

> This is effect can be observed with Open Source as well, you can read more about it [here](https://doi.org/10.1109/MS.2013.95). 

Reviewer 2: Should we link to a document that is not publicly-available?

Reviewer 2: 🆗
